### PR TITLE
Make name collision check more strict

### PR DESCRIPTION
**Pull request by cocreature**
This PR extends the name collision check to catch collisions between
A:B (type B in module A) and module A.B.C. For now this is just a
warning and not an error. Once we turn it into an error, we also need
to add this to the Scala collision checker.

There is a fair bit of plumbing required to make warnings work but on
the plus side we get multiple errors at once now instead of erroring
out on the first one.

changelog_begin

- [DAML Compiler] The name collision check has been extended to also
  count the case as a collision where you have a type B in module A and a module
  A.B.C (but no module A.B). This is a warning in this SDK release but
  will become an error in a future release. The typescript codegen is
  not usable on packages that don’t uphold this restriction.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker.hs [hunk 1]
```diff
@@ -20,36 +19,39 @@ import qualified DA.Daml.LF.TypeChecker.PartyLiterals as PartyLits
...
-      Recursion.checkModule m
-      Check.checkModule m
-      Serializability.checkModule m
-      PartyLits.checkModule m
-    NameCollision.runCheckModuleDeps world m
+checkModuleInWorld :: World -> Version -> Module -> [Diagnostic]
+checkModuleInWorld world version m =
+    case typeCheckResult of
+        Left err -> [toDiagnostic DsError err]
+        Right () -> NameCollision.runCheckModuleDeps world m
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker.hs]: Would it make sense to deliver the errors and the warnings in one go, i.e., append the diagnostics rather than branch here?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs [hunk 2]
```diff
@@ -7,11 +7,14 @@ module DA.Daml.LF.TypeChecker.Error(
     TemplatePart(..),
     UnserializabilityReason(..),
     SerializabilityRequirement(..),
-    errorLocation
+    errorLocation,
+    toDiagnostic
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs]: Haskell allows trailing commas here. I like that since it helps fighting future diff noise. 😃

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs [hunk 3]
```diff
@@ -107,41 +128,49 @@ fullyResolve = FRName . map T.toLower . \case
 -- | State of the name collision checker. This is a
 -- map from fully resolved names within a package to their
 -- original names. We update this map as we go along.
-newtype NCState = NCState (M.Map FRName [Name])
+data NCState = NCState (M.Map FRName [Name])
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs]: What's wrong with the `newtype`? Not lazy enough?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs [hunk 4]
```diff
@@ -107,41 +128,49 @@ fullyResolve = FRName . map T.toLower . \case
...
+    | null badNames =
         Right . NCState $ M.insert frName (name : oldNames) nameMap
-    else do
-        Left $ EForbiddenNameCollision
-            (displayName name)
-            (map displayName badNames)
+    | otherwise =
+        let err = EForbiddenNameCollision (displayName name) (map displayName badNames)
+            diag = toDiagnostic DsError err
+        -- If name is virtual are all badNames are virtual, we demote it to a
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs]: According to the code, you probably mean:
```suggestion
        -- If name is virtual or all badNames are virtual, we demote it to a
```

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs [hunk 5]
```diff
@@ -222,8 +257,15 @@ checkModuleDeps world mod0 = do
...
-runCheckPackage :: Package -> Either Error ()
+runCheckPackage :: Package -> [Diagnostic]
 runCheckPackage = runNameCollision . checkPackage
+
+-- | Given module name A.B.C produce the virtual module names A and A.B.
+virtualModuleNames :: Module -> [(ModuleName, ModuleName)]
+virtualModuleNames Module{moduleName = origin@(ModuleName components)}
+    | null components = error "Empty module names are invalid."
+    | otherwise = map (\v -> (ModuleName v, origin))  ((tail . inits . init) components)
+    -- init goes from A.B.C to A.B, inits gives us [[], [A], [A,B]], tail drops []
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs]: I had exactly the same chain of functions in my prototype. 😃

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 6]
```diff
@@ -352,6 +353,31 @@ tests tools@Tools{damlc} = testGroup "Packaging" $
...
+        writeFileUTF8 (projDir </> "src" </> "A" </> "B" </> "C.daml") $ unlines
+            [ "module A.B.C where"
+            , "data C = C Int"
+            ]
+        (exitCode, out, err) <- readProcessWithExitCode damlc ["build", "--project-root", projDir] ""
+        out @?= unlines
+            [ "Compiling proj to a DAR."
+            , "Created .daml/dist/proj-0.0.1.dar"
+            ]
+        assertInfixOf "collision between variant A:B and virtual module A.B (from A.B.C)" err
```
Comment (hurryabit) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: What do you think of speaking about "module name prefixes" in the error messages?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs [hunk 7]
```diff
@@ -24,6 +25,13 @@ import Control.Monad.Except (throwError)
 -- see 'FRName'.
 data Name
     = NModule ModuleName
+    | NVirtualModule ModuleName ModuleName
+    -- ^ For a module A.B.C we insert virtual module names A and A.B.
+    -- This is an extension of the check that prevents you from having
+    -- a type B in module A if you have a module A.B to also fail if you
+    -- have a module A.B.C but not A.B. At the moment, we only emit a warning
```
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs]: I completely misunderstood this comment when first reading. I read it as: We want to enforce that `A.B` exists whenever `A.B.C` exists. I suggest we slightly reword this to something like:

> This is used to check that you do not have a type `B` in a module `A` and a module `A.B.C` at the same time, even if you don't have a module `A.B`.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker.hs (hunk 1 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker.hs]: Changed, not sure how useful it is in practice given that things are pretty bad if the LF typechecker fails but it doesn’t do any harm.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs (hunk 2 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs]: Fair, I tend to find it easier to never use them in Haskell than having to remember the one place where they are allowed but I’ve changed it now :slightly_smiling_face:

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs (hunk 3 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs]: oops, that was leftover from an experiment where I tried to add two fields to this, reverted.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs (hunk 4 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs]: english is hard, thanks :)

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 6 above)
Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: Changed it to use `module prefix` instead of `virtual module` in the error message. I still call it virtual module internally since I think that makes the code easier to understand since it treats them pretty much exactly like other module names.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs (hunk 7 above)
Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs]: Thanks, copied your wording!

File: compiler/damlc/tests/src/DA/Test/Packaging.hs (hunk 6 above)
Comment (hurryabit) [reply compiler/damlc/tests/src/DA/Test/Packaging.hs]: I totally agree regarding the internal naming.
